File: test/e2e/Apps/BasicNode/src/functions/ActivityInputType.ts
// Copyright (c) .NET Foundation. All rights reserved.
// Licensed under the MIT License. See License.txt in the project root for license information.

import * as df from 'durable-functions';
import { OrchestrationHandler, OrchestrationContext, ActivityHandler } from "durable-functions";

// Helper for durations
function parseDuration(duration: string): number {
    // Parses "HH:MM:SS" or "0HH:MM:SS" to milliseconds
    const match = duration.match(/(\d+):(\d+):(\d+)/);
    if (!match) return 0;
    const [, h, m, s] = match.map(Number);
    return ((h || 0) * 3600 + (m || 0) * 60 + (s || 0)) * 1000;
}

class MyCustomClass {
    name?: string;
    age: number;
    data: number[];
    duration: number; // milliseconds

    constructor(name: string | undefined, age: number, data: number[], duration: number) {
        this.name = name;
        this.age = age;
        this.data = data;
        this.duration = duration;
    }

    toString(): string {
        // Leading 0 before duration to match expected output
        const durationStr = `${new Date(this.duration).toISOString().substring(11, 19)}`;
        return `{Name: ${this.name}, Age: ${this.age}, Duration: ${durationStr}, Data: [${Array.from(this.data).join(", ")}]}`;
    }

    // toJSON(): object {
    //     return {
    //         Name: this.name,
    //         Age: this.age,
    //         Data: Array.from(this.data),
    //         Duration: `0${new Date(this.duration).toISOString().substr(11, 8)}`
    //     };
    // }

    static fromJSON(data: any): MyCustomClass {
        return new MyCustomClass(
            data.name,
            data.age,
            Array.from(data.data),
            data.duration
        );
    }
}

// Orchestrator
const ActivityInputTypeOrchestrator: OrchestrationHandler = function* (context: OrchestrationContext) {
    const output: string[] = [];

    // Test byte array input
    const byte_array_input = [1, 2, 3, 4, 5];
    const r_1 = yield context.df.callActivity("byte_array_input", byte_array_input);
    output.push(r_1);

    // Test empty byte array input
    const empty_byte_array: number[] = [];
    const r_2 = yield context.df.callActivity("byte_array_input", empty_byte_array);
    output.push(r_2);

    // Test single byte input
    const single_byte_input = 42;
    const r_3 = yield context.df.callActivity("single_byte_input", single_byte_input);
    output.push(r_3);

    // Test custom class input
    const custom_class_input = new MyCustomClass("Test", 25, [1, 2, 3], 3600000); // 1 hour in ms
    const r_4 = yield context.df.callActivity("custom_class_input", custom_class_input);
    output.push(r_4);

    // Test int array input
    const int_array_input = [1, 2, 3, 4, 5];
    const r_5 = yield context.df.callActivity("int_array_input", int_array_input);
    output.push(r_5);

    // Test string input
    const string_input = "Test string input";
    const r_6 = yield context.df.callActivity("string_input", string_input);
    output.push(r_6);

    // Test array of custom class input
    const complex_input = [
        new MyCustomClass("Test1", 25, [1, 2, 3], 30 * 60 * 1000), // 30 min
        new MyCustomClass("Test2", 30, [], 45 * 60 * 1000) // 45 min
    ];
    const r_7 = yield context.df.callActivity("custom_class_array_input", complex_input);
    output.push(r_7);

    return output;
};
df.app.orchestration("ActivityInputTypeOrchestrator", ActivityInputTypeOrchestrator);

// Activities

const byte_array_input: ActivityHandler = (input: Uint8Array): string => {
    if (!Array.isArray(input) || !input.every(x => typeof x === "number")) {
        return `Error: Expected byte[] but got ${typeof input} ${input}`;
    }
    return `Received byte[]: [${Array.from(new Uint8Array(input)).join(", ")}]`;
};
df.app.activity("byte_array_input", { handler: byte_array_input });

const single_byte_input: ActivityHandler = (input: number): string => {
    if (typeof input !== "number") {
        return `Error: Expected byte but got ${typeof input}`;
    }
    return `Received byte: ${input}`;
};
df.app.activity("single_byte_input", { handler: single_byte_input });

const custom_class_input: ActivityHandler = (input: MyCustomClass): string => {
    const obj = MyCustomClass.fromJSON(input);
    const data = obj.data;
    if (!(Array.isArray(data) && data.every(x => typeof x === "number"))) {
        return `Error: Expected Data to be byte[] but got ${typeof data}`;
    }
    return `Received CustomClass: ${obj.toString()}`;
};
df.app.activity("custom_class_input", { handler: custom_class_input });

const int_array_input: ActivityHandler = (input: number[]): string => {
    if (!Array.isArray(input) || !input.every(x => typeof x === "number")) {
        return `Error: Expected int[] but got ${typeof input}`;
    }
    return `Received int[]: [${input.join(", ")}]`;
};
df.app.activity("int_array_input", { handler: int_array_input });

const string_input: ActivityHandler = (input: string): string => {
    if (typeof input !== "string") {
        return `Error: Expected string but got ${typeof input}`;
    }
    return `Received string: ${input}`;
};
df.app.activity("string_input", { handler: string_input });

const custom_class_array_input: ActivityHandler = (input: any): string => {
    let parsedInput: MyCustomClass[] = [];
    if (!Array.isArray(input)) {
        return `Error: Expected MyCustomClass[] but got ${typeof input}`;
    }
    for (const item of input) {
        if (!(item instanceof MyCustomClass) &&
            !(item && typeof item === "object" && "name" in item && "age" in item && "data" in item && "duration" in item)) {
            return `Error: Expected MyCustomClass but got ${typeof item}`;
        }
        parsedInput.push(item instanceof MyCustomClass ? item : MyCustomClass.fromJSON(item));
    }
    return `Received CustomClass[]: [${Array.from(parsedInput).map(x => x.toString()).join(", ")}]`;
};
df.app.activity("custom_class_array_input", { handler: custom_class_array_input });

File: test/e2e/Apps/BasicJava/src/main/java/com/function/ActivityInputType.java
package com.function;

import com.microsoft.azure.functions.annotation.*;
import com.microsoft.azure.functions.*;
import com.microsoft.durabletask.*;
import com.microsoft.durabletask.azurefunctions.DurableActivityTrigger;
import com.microsoft.durabletask.azurefunctions.DurableOrchestrationTrigger;

import java.util.*;
import java.time.Duration;


/**
 * Durable Functions E2E test for activity input types.
 */
public class ActivityInputType {

    /**
     * Orchestrator function that tests various activity input types.
     */
    @FunctionName("ActivityInputTypeOrchestrator")
    public List<String> activityInputTypeOrchestrator(
            @DurableOrchestrationTrigger(name = "context") TaskOrchestrationContext context) {
        List<String> output = new ArrayList<>();

        // Test byte array input
        List<Byte> byteArrayInput = Arrays.asList((byte) 1, (byte) 2, (byte) 3, (byte) 4, (byte) 5);
        output.add(context.callActivity("ByteArrayInput", byteArrayInput, String.class).await());

        // Test empty byte array input
        List<Byte> emptyByteArray = new ArrayList<>();
        output.add(context.callActivity("ByteArrayInput", emptyByteArray, String.class).await());

        // Test single byte input
        byte singleByteInput = 42;
        output.add(context.callActivity("SingleByteInput", singleByteInput, String.class).await());

        // Test custom class input
        CustomClass customClassInput = new CustomClass();
        customClassInput.setName("Test");
        customClassInput.setAge(25);
        customClassInput.setData(Arrays.asList((byte) 1, (byte) 2, (byte) 3));
        customClassInput.setDuration(Duration.ofHours(1).toMillis());
        output.add(context.callActivity("CustomClassInput", customClassInput, String.class).await());

        // Test int array input
        int[] intArrayInput = new int[] { 1, 2, 3, 4, 5 };
        output.add(context.callActivity("IntArrayInput", intArrayInput, String.class).await());

        // Test string input
        String stringInput = "Test string input";
        output.add(context.callActivity("StringInput", stringInput, String.class).await());

        // Test array of custom class input
        List<CustomClass> complexInput = Arrays.asList(
            new CustomClass("Test1", 25, Arrays.asList((byte) 1, (byte) 2, (byte) 3), Duration.ofMinutes(30)),
            new CustomClass("Test2", 30, new ArrayList<>(), Duration.ofMinutes(45))
        );
        output.add(context.callActivity("CustomClassArrayInput", complexInput, String.class).await());

        return output;
    }

    /**
     * Activity: Receives a byte array.
     */
    @FunctionName("ByteArrayInput")
    public String byteArrayInput(
            @DurableActivityTrigger(name = "input") List<Byte> input,
            final ExecutionContext context) {
        return "Received byte[]: " + input;
    }

    /**
     * Activity: Receives a single byte.
     */
    @FunctionName("SingleByteInput")
    public String singleByteInput(
            @DurableActivityTrigger(name = "input") byte input,
            final ExecutionContext context) {
        return "Received byte: " + input;
    }

    /**
     * Activity: Receives a custom class.
     */
    @FunctionName("CustomClassInput")
    public String customClassInput(
            @DurableActivityTrigger(name = "input") CustomClass input,
            final ExecutionContext context) {
        if (input.getData() == null || !(input.getData() instanceof List<Byte>)) {
            return "Error: Expected Data to be byte[] but got " + (input.getData() == null ? "null" : input.getData().getClass().getSimpleName());
        }
        return "Received CustomClass: " + input.toString();
    }

    /**
     * Activity: Receives an int array.
     */
    @FunctionName("IntArrayInput")
    public String intArrayInput(
            @DurableActivityTrigger(name = "input") int[] input,
            final ExecutionContext context) {
        return "Received int[]: [" + joinInts(input) + "]";
    }

    /**
     * Activity: Receives a string.
     */
    @FunctionName("StringInput")
    public String stringInput(
            @DurableActivityTrigger(name = "input") String input,
            final ExecutionContext context) {
        input = input.substring(1, input.length() - 1); // Bug: Double serialization issue
        return "Received string: " + input;
    }

    /**
     * Activity: Receives an array of custom class.
     */
    @FunctionName("CustomClassArrayInput")
    public String customClassArrayInput(
            @DurableActivityTrigger(name = "input") List<CustomClass> input,
            final ExecutionContext context) {
        for (CustomClass item : input) {
            if (item.getData() == null || !(item.getData() instanceof List<Byte>)) {
                return "Error: Expected Data to be byte[] but got " + (item.getData() == null ? "null" : item.getData().getClass().getSimpleName());
            }
        }
        return "Received CustomClass[]: " + input.toString();
    }

    // Helper to join int arrays as comma-separated string
    private static String joinInts(int[] arr) {
        if (arr == null || arr.length == 0) return "";
        StringBuilder sb = new StringBuilder();
        for (int i = 0; i < arr.length; i++) {
            if (i > 0) sb.append(", ");
            sb.append(arr[i]);
        }
        return sb.toString();
    }
}

/**
 * Custom class for activity input.
 */
class CustomClass {
    private String name;
    private int age;
    private List<Byte> data;
    // As of Java 16, reflective access to Duration (and all Java base modules) was restricted,
    // so serialization of objects with a Duration field is not possible directly. Due to this, we
    // store the duration as milliseconds for seralization/deserialization.
    private long duration;

    public CustomClass() {}

    public CustomClass(String name, int age, List<Byte> data, Duration duration) {
        this.name = name;
        this.age = age;
        this.data = data;
        this.duration = duration == null ? 0 : duration.toMillis();
    }

    public String getName() { return name; }
    public void setName(String name) { this.name = name; }

    public int getAge() { return age; }
    public void setAge(int age) { this.age = age; }

    public List<Byte> getData() { return data; }
    public void setData(List<Byte> data) { this.data = data; }

    public long getDuration() { return duration; }
    public void setDuration(long duration) { this.duration = duration; }
    
    public String toString() {
        return String.format("{Name: %s, Age: %d, Duration: %s, Data: %s}",
            name, age, formatDuration(), data);
    }

    private String formatDuration() {
        Duration duration = Duration.ofMillis(this.duration);

        long hours = duration.toHours();
        long minutes = duration.toMinutesPart();
        long seconds = duration.toSecondsPart();

        String formatted = String.format("%02d:%02d:%02d", hours, minutes, seconds);
        return formatted;
    }
}